api/src/trace/SugaredTracer.ts

```diff
@@ -0,0 +1,210 @@
+   *     });
+   * @example
+   *     const something = await tracer.withActiveSpan('op', span => {
+   *      // do some async work
+   *     });
+   */
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    context: Context,
+    fn: F,
+  ): ReturnType<F>
+  withActiveSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    arg2: SugaredSpanOptions | F,
+    arg3?: Context | F,
+    arg4?: F,
+  ): ReturnType<F> {
+    const {opts, ctx, fn} = massageParams(arg2, arg3, arg4);
+
+    return this._tracer.startActiveSpan(name, opts, ctx,
+      (span: Span) => handleFn(span, opts, fn)) as ReturnType<F>;
+  }
+
+  /**
+   * Starts a new {@link Span} and ends it after execution of fn without setting it on context.
+   * The span will be closed after the function has executed.
+   * If an exception occurs, it is recorded, the status is et to ERROR and rethrown.
+   *
+   * This method do NOT modify the current Context.
+   *
+   * @param name The name of the span
+   * @param [options] SugaredSpanOptions used for span creation
+   * @param [context] Context to use to extract parent
+   * @param fn function called in the context of the span and receives the newly created span as an argument
+   * @returns Span The newly created span
+   * @example
+   *     const something = tracer.withSpan('op', span => {
+   *      // do some work
+   *     });
+   * @example
+   *     const something = await tracer.withSpan('op', span => {
+   *      // do some async work
+   *     });
+   */
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    context: Context,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    context: Context,
+    fn: F,
+  ): ReturnType<F>
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    arg2: SugaredSpanOptions | F,
+    arg3?: Context | F,
+    arg4?: F,
+  ): ReturnType<F> {
+    const {opts, ctx, fn} = massageParams(arg2, arg3, arg4);
+
+    const span = this._tracer.startSpan(name, opts, ctx);
+    return handleFn(span, opts, fn) as ReturnType<F>;
+  }
+}
+
+/**
+ * Massages parameters of withSpan and withActiveSpan to allow signature overwrites
+ * @param arg
+ * @param arg2
+ * @param arg3
+ */
+function massageParams<F extends (span: Span) => ReturnType<F>>(arg: SugaredSpanOptions | F,
```

Comment:
I can't think of a good name for this but the existing name doesn't describe at all what it does. Maybe something like `getVariadicParams`?

Reply:
IMO the `get` suggests that we are retrieving the params somewhere from. What do you think about `processParams` or `normalizeParams`?